File: backend/api/v1/endpoints/extract_audio.py
from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
import os
import sys
import subprocess
from typing import Dict, Any

from backend.db.session import get_db
from backend.api import deps
from backend.db.models import Capture

router = APIRouter()

@router.post("/{capture_id}", response_model=Dict[str, Any])
def extract_audio_for_capture(
    capture_id: int,
    db: Session = Depends(deps.get_db),
    current_user = Depends(deps.get_current_active_user),
):
    """
    Extract audio for a specific capture ID.
    This endpoint runs the extract_audio_for_capture.py script for a specific capture ID.
    """
    try:
        # Path to the extract_audio_for_capture.py script
        script_path = os.path.join("/app", "scripts", "extract_audio_for_capture.py")
        
        if not os.path.exists(script_path):
            raise HTTPException(status_code=404, detail=f"Script not found at {script_path}")
        
        # Run the script in a non-blocking way
        cmd = [sys.executable, script_path, str(capture_id)]
        subprocess.Popen(cmd)
        
        return {
            "success": True,
            "message": f"Audio extraction started for capture ID: {capture_id}"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error extracting audio: {str(e)}")

File: frontend/utils/extractAudio.ts
import { apiClient } from './apiClient';

interface ExtractAudioResponse {
  success: boolean;
  message: string;
  error?: string;
  output_file?: string;
}

/**
 * Extract audio for a specific capture ID
 * @param captureId The ID of the capture to extract audio for
 * @returns Promise that resolves with the extraction result
 */
export const extractAudioForCapture = async (captureId: number): Promise<ExtractAudioResponse> => {
  try {
    const response = await apiClient.post<ExtractAudioResponse>(`/parliament-tv/audio-extraction/${captureId}`);
    console.log('Audio extraction response:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('Error extracting audio:', error);
    
    // Try to extract error information from the response if available
    let errorMessage = 'Failed to extract audio';
    if (error.response && error.response.data) {
      if (error.response.data.error) {
        errorMessage = error.response.data.error;
      } else if (error.response.data.detail) {
        errorMessage = error.response.data.detail;
      }
    }
    
    return {
      success: false,
      message: errorMessage,
      error: errorMessage
    };
  }
};
